src/meta/src/barrier/progress.rs

```diff
@@ -122,7 +199,54 @@ impl<S: MetaStore> CreateMviewProgressTracker<S> {
             self.actor_map.insert(actor, ddl_epoch);
         }
 
-        let progress = Progress::new(actors);
+        let (creating_mv_id, upstream_mv_count, upstream_total_key_count, definition) =
+            if let Command::CreateStreamingJob {
+                table_fragments,
+                dispatchers,
+                table_mview_map,
+                definition,
+                ..
+            } = &command.context.command
+            {
+                // Keep track of how many times each upstream MV appears.
+                let mut upstream_mv_count = HashMap::new();
+                for (table_id, actors) in table_mview_map {
+                    assert!(!actors.is_empty());
+                    let dispatch_count: usize = dispatchers
+                        .iter()
+                        .filter(|(upstream_actor_id, _)| actors.contains(upstream_actor_id))
+                        .map(|(_, v)| v.len())
+                        .sum();
+                    upstream_mv_count.insert(*table_id, dispatch_count / actors.len());
+                }
+
+                let upstream_total_key_count: u64 = upstream_mv_count
+                    .iter()
+                    .map(|(upstream_mv, count)| {
+                        *count as u64
+                            * version_stats
+                                .table_stats
+                                .get(&upstream_mv.table_id)
+                                .map_or(0, |stat| stat.total_key_count as u64)
```

Comment:
Hummock's stats are for the total keys, including the historical ones. Assuming that there're a lot of `Updates` in the upstream mview, if the rows come from the "upstream", we'll count them correctly, but for those from "snapshot", a single consumed row may correspond to multiple historical rows in the stats. Not sure whether this impacts a lot.

Reply:
Yes, our progress calculation is an approximate value instead of accurate value. It is hard to calculate the accurate value based on what we have currently. From the user perspective, I think they will care about the DDL will take a long time that is the snapshot contains a large amount of data. In this case, I hope with the help of compaction, we can guarantee the total key count will not exceed (for example) 30% of the actual key count. BTW, if total key count is larger than the actual key count, our calculated progress number is smaller than the actual progress and this is acceptable as long as we provide a lower bound value.